lib/charms/tempo_k8s/v0/snapshot.py

```diff
@@ -0,0 +1,321 @@
+"""Snapshot charm library.
+
+This library acts as a bridge to scenario's
+[State](https://github.com/canonical/ops-scenario/blob/main/scenario/state.py#L861) data structure,
+allowing you to capture snapshots of a live charm's current state and serialize them as json.
+"""
+import logging
+from dataclasses import asdict, fields, is_dataclass
+from datetime import datetime
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Dict, List, cast
+
+import ops
+import scenario
+import scenario.runtime
+from ops import SecretRotate, pebble
+from scenario.state import _EntityStatus
+
+if TYPE_CHECKING:
+    from scenario.state import AnyRelation
+
+logger = logging.getLogger("snapshot")
+
+# The unique Charmhub library identifier, never change it
+LIBID = "012181337a3a4717861b5fb4d8bdfc2b"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = ["ops-scenario>=6"]
+
+
+class SnapshotBackend:
```

Comment:
In python 4, classes should have `.__json__` :pray: 
Could pydantic help in the meanwhile?
Otherwise it seems like the raw output of the hook tools would serve you better here.

Reply:
Iirc, hook tools already return json, so by calling the _ModelBackend hook tools methods directly you may not need to do much else.

Reply:
mm, you're saying this because you want to avoid the scenario dependency I think, and I think it's a fair point. And by bypassing the framework too, we can also avoid having to serialize an ops.Relation.
Okay, I'm on board.